File: BSD3203_CAT2_NziokaEmmanuelMunyao2004860.r
# BSD 3203 CAT 2 FOR 20/04860 NZIOKA EMMANUEL MUNYAO(R)
# The code of this CAT, below, was written and edited on Notepad++ v8.7.8 64 bit and tested and run on R 4.4.2 GUI 64 bit.
# All the steps are highlighted and labelled correctly
# The machine used for this CAT  supports displaying date and time in the long format however, 
#  R code was used to display the current date and time for every question heading. The Machine name is Emmanuel-Rogger(Emmanuel Nzioka).
# In case of any problems, the author of this script can be reached through +254791971305 or 2004860@students.ac.ke
										# Section A: R Programming (20 Marks)
										# Question 1: Data Manipulation with dplyr (6 Marks)
# Get current date and time
current_datetime <- Sys.time()
# Get machine name
machine_name <- Sys.info()["nodename"]
# Print the results
print(paste("Current Date and Time:", current_datetime))
print(paste("Machine Name:", machine_name))
# Load necessary libraries
library(dplyr)
# Create sales_data dataframe
sales_data <- data.frame(
  Product = c("Laptop", "Phone", "Tablet", "Laptop", "Phone"),
  Region = c("Region A", "Region B", "Region A", "Region B", "Region A"),
  Sales = c(7000, 4500, 8000, 3000, 6500),
  Date = as.Date(c("2024-01-10", "2024-01-12", "2024-01-15", "2024-01-18", "2024-01-20"))
)
# Save dataset to CSV
write.csv(sales_data, "sales_data.csv", row.names = FALSE)
# Load dataset from CSV
sales_data <- read.csv("sales_data.csv")
# Display only Product and Sales columns
sales_data %>% select(Product, Sales)
# Filter rows where Sales is greater than 5000
sales_data %>% filter(Sales > 5000)
# Group data by Region and calculate total sales per region
region_sales <- sales_data %>%
  group_by(Region) %>%
  summarise(Total_Sales = sum(Sales))
head(region_sales)
# Save to CSV
write.csv(region_sales, "region_sales.csv", row.names = FALSE)
										# Question 2: Data Visualization with ggplot2 (6 Marks)
# Get current date and time
current_datetime <- Sys.time()
# Get machine name
machine_name <- Sys.info()["nodename"]
# Print the results
print(paste("Current Date and Time:", current_datetime))
print(paste("Machine Name:", machine_name))
# Load necessary library
library(ggplot2)
# Create a bar chart showing total sales for each Product
ggplot(sales_data, aes(x = Product, y = Sales, fill = Product)) +
  geom_bar(stat = "identity") +
  ggtitle("Total Sales per Product") +
  theme_minimal()
# Create a line plot showing sales trends over time for Region A
sales_data$Date <- as.Date(sales_data$Date)  # Convert Date column to Date format
region_a_sales <- filter(sales_data, Region == "Region A")
ggplot(region_a_sales, aes(x = Date, y = Sales, group = 1)) +
  geom_line(color = "blue") +
  geom_point() +
  ggtitle("Sales Trend Over Time for Region A") +
  theme_minimal()
											# Question 3: Statistical Analysis (8 Marks)
# Get current date and time
current_datetime <- Sys.time()
# Get machine name
machine_name <- Sys.info()["nodename"]
# Print the results
print(paste("Current Date and Time:", current_datetime))
print(paste("Machine Name:", machine_name))
# Summary statistics for Sales column
summary(sales_data$Sales)
# Perform t-test comparing mean Sales between Region A and Region B
sales_region_a <- filter(sales_data, Region == "Region A")$Sales
sales_region_b <- filter(sales_data, Region == "Region B")$Sales
t_test_result <- t.test(sales_region_a, sales_region_b, var.equal = TRUE)
t_test_result
# Interpretation of p-value
if (t_test_result$p.value < 0.05) {
  print("The difference in mean sales between Region A and Region B is statistically significant.")
} else {
  print("The difference in mean sales between Region A and Region B is not statistically significant.")
}

File: BSD3203_CAT2_NziokaEmmanuelMunyao2004860.py
# BSD 3203 CAT 2 FOR 20/04860 NZIOKA EMMANUEL MUNYAO(PYTHON)
# The code of this CAT, below, was written and edited on Notepad++ v8.7.8 64 bit and tested and run on Google Colab
# All the steps are highlighted and labelled correctly
# The machine used for this CAT  supports displaying date and time in the long format however, 
#  Python code was used to display the current date and time for every question heading. The Machine name is Emmanuel-Rogger(Emmanuel Nzioka).
# In case of any problems, the author of this script can be reached through +254791971305 or 2004860@students.ac.ke
                            # Section B: Python Programming (20 Marks)
                            # Question 4: Data Manipulation with Pandas (6 Marks)
import datetime
# Get current date and time
current_datetime = datetime.datetime.now()
# Print results
print("Current Date and Time:", current_datetime)
import pandas as pd
# Create employee_data dataframe
employee_data = pd.DataFrame({
    "Name": ["John Doe", "Jane Smith", "Mike Brown", "Anna White", "Emma Davis"],
    "Department": ["HR", "IT", "Sales", "IT", "Sales"],
    "Salary": [55000, 75000, 48000, 82000, 61000],
    "Hire_Date": pd.to_datetime(["2020-06-15", "2018-03-22", "2019-10-10", "2021-01-05", "2017-07-19"])
})
# Save dataset to CSV
employee_data.to_csv("employee_data.csv", index=False)
# Load dataset from CSV
employee_data = pd.read_csv("employee_data.csv")
# Display only Name and Salary columns
print(employee_data[["Name", "Salary"]])
# Filter records where Salary is above 60000
print(employee_data[employee_data["Salary"] > 60000])
# Group data by Department and calculate the average salary per department
department_salary = employee_data.groupby("Department")["Salary"].mean().reset_index()
# Save to CSV
department_salary.to_csv("department_salary.csv", index=False)
                                        # Question 5: Data Visualization with Matplotlib / Seaborn (6 Marks)
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
# Get current date and time
current_datetime = datetime.datetime.now()
# Print results
print("Current Date and Time:", current_datetime)
# Histogram of Salary distribution
plt.figure(figsize=(8,5))
sns.histplot(employee_data["Salary"], bins=5, kde=True)
plt.title("Salary Distribution")
plt.xlabel("Salary")
plt.ylabel("Frequency")
plt.show()
# Boxplot showing Salary by Department
plt.figure(figsize=(8,5))
sns.boxplot(x="Department", y="Salary", data=employee_data)
plt.title("Salary by Department")
plt.show()
                                            # Question 6: Machine Learning - Linear Regression (8 Marks)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import datetime
# Get current date and time
current_datetime = datetime.datetime.now()
# Print results
print("Current Date and Time:", current_datetime)
# Create house_prices dataframe
house_prices = pd.DataFrame({
    "Square_Feet": [1500, 1800, 2400, 3000, 1200],
    "Bedrooms": [3, 4, 4, 5, 2],
    "Price": [300000, 350000, 450000, 600000, 250000]
})
# Save dataset to CSV
house_prices.to_csv("house_prices.csv", index=False)
# Load dataset from CSV
house_prices = pd.read_csv("house_prices.csv")
# Define features and target variable
X = house_prices[["Square_Feet", "Bedrooms"]]
y = house_prices["Price"]
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Create Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)
# Display model coefficients and intercept
print("Coefficients:", model.coef_)
print("Intercept:", model.intercept_)
# Predict the price of a house with 2000 square feet and 3 bedrooms
predicted_price = model.predict([[2000, 3]])
print("Predicted price for a 2000 sq ft, 3-bedroom house:", predicted_price[0])